File: android/app/src/main/java/com/pakrinha/khmerphonetic/Suggestions.kt
package com.pakrinha.khmerphonetic

/**
 * The word suggestions that make this keyboard worth using: you type roughly how
 * a word sounds and it finds the Khmer, the way a Pinyin IME does.
 *
 * Two lexicons, searched in tiers, ported from dict.js:
 *
 *   curated.tsv  ~440 common words, each listed under every romanization the web
 *                app's reverse-transliterator derives for it. Those keys are
 *                precomputed by tools/build_android.mjs rather than regenerated
 *                here, so they cannot come out different from the browser's.
 *   words.txt    62,000 words from the Google Khmer pronunciation lexicon,
 *                stored as coarse sound-class skeletons and ordered by corpus
 *                frequency, so line number is rank.
 *
 * Tiers, lowest wins: personal(-1), curated exact(0), curated prefix(1),
 * curated vowel-skeleton(2, 3), lexicon skeleton exact(4), prefix(5).
 */
class Suggestions(
    curatedLines: Sequence<String>,
    lexiconLines: Sequence<String>,
    private val learned: Learning = Learning.NONE,
) {

    data class Candidate(val khmer: String, val gloss: String = "")

    private class Curated(val key: String, val vskel: String, val khmer: String, val gloss: String)

    private val curated = ArrayList<Curated>(16000)
    private val lexKhmer = ArrayList<String>(62000)
    private val lexSkel = ArrayList<String>(62000)

    /* Sorted views used to binary-search a prefix range instead of scanning.
     *
     * The first version walked all ~77,000 entries on every keystroke. On a
     * desktop JVM that is 0.6 ms and looks fine; on a phone it is tens of
     * milliseconds of main-thread work per letter, which both delays the word
     * bar and makes the keyboard miss taps. Each array below holds row indices
     * ordered by the string being searched, so a lookup touches only the
     * matching range. */
    private lateinit var curatedByKey: IntArray
    private lateinit var curatedByVskel: IntArray
    private lateinit var lexBySkel: IntArray

    init {
        for (line in curatedLines) {
            if (line.isBlank() || line.startsWith("#")) continue
            val p = line.split('\t')
            if (p.size >= 3) curated.add(Curated(p[0], p[1], p[2], if (p.size > 3) p[3] else ""))
        }
        for (line in lexiconLines) {
            if (line.isBlank() || line.startsWith("#")) continue
            val tab = line.indexOf('\t')
            if (tab > 0) {
                lexKhmer.add(line.substring(0, tab))
                lexSkel.add(line.substring(tab + 1).trim())
            }
        }
        curatedByKey = sortedIndices(curated.size) { curated[it].key }
        curatedByVskel = sortedIndices(curated.size) { curated[it].vskel }
        lexBySkel = sortedIndices(lexSkel.size) { lexSkel[it] }
    }

    val lexiconSize: Int get() = lexKhmer.size

    private inline fun sortedIndices(size: Int, crossinline key: (Int) -> String): IntArray =
        (0 until size).sortedBy(key).toIntArray()

    /** First position in [order] whose string is >= [target]. */
    private inline fun lowerBound(order: IntArray, target: String, crossinline key: (Int) -> String): Int {
        var lo = 0
        var hi = order.size
        while (lo < hi) {
            val mid = (lo + hi) ushr 1
            if (key(order[mid]) < target) lo = mid + 1 else hi = mid
        }
        return lo
    }

    /**
     * Visit every row whose string starts with [prefix], in sorted order.
     * [visit] returns false to stop early.
     */
    private inline fun forEachWithPrefix(
        order: IntArray,
        prefix: String,
        crossinline key: (Int) -> String,
        visit: (row: Int, value: String) -> Boolean,
    ) {
        var i = lowerBound(order, prefix, key)
        while (i < order.size) {
            val row = order[i]
            val value = key(row)
            if (!value.startsWith(prefix)) return
            if (!visit(row, value)) return
            i++
        }
    }

    /** What the user has picked before, so their words float up. */
    interface Learning {
        fun uses(khmer: String): Int
        fun personal(): Map<String, String>

        companion object {
            val NONE = object : Learning {
                override fun uses(khmer: String) = 0
                override fun personal(): Map<String, String> = emptyMap()
            }
        }
    }

    private class Hit(var tier: Int, var sub: Long, var gloss: String)

    fun suggest(rawQuery: String, limit: Int = 8): List<Candidate> {
        var q = rawQuery.lowercase().trim()
        if (q.isEmpty()) return emptyList()
        // Bound the work: a long paste should never make the keyboard hang.
        if (q.length > MAX_TOKEN) q = q.substring(q.length - MAX_TOKEN)

        val hits = LinkedHashMap<String, Hit>()
        fun add(khmer: String, tier: Int, sub: Long, gloss: String = "") {
            val existing = hits[khmer]
            if (existing == null) {
                hits[khmer] = Hit(tier, sub, gloss)
                return
            }
            if (tier < existing.tier || (tier == existing.tier && sub < existing.sub)) {
                existing.tier = tier
                existing.sub = sub
            }
            if (existing.gloss.isEmpty() && gloss.isNotEmpty()) existing.gloss = gloss
        }

        for ((spelling, khmer) in learned.personal()) {
            if (spelling == q) add(khmer, -1, 0, GLOSS_YOURS)
            else if (spelling.startsWith(q)) add(khmer, -1, (spelling.length - q.length).toLong(), GLOSS_YOURS)
        }

        val qv = vskel(q)
        val fuzzyOk = q.length >= 3

        // Spelled exactly, or the query is the start of a longer spelling.
        forEachWithPrefix(curatedByKey, q, { curated[it].key }) { row, key ->
            val c = curated[row]
            if (key == q) add(c.khmer, 0, 0, c.gloss)
            else add(c.khmer, 1, (key.length - q.length).toLong(), c.gloss)
            true
        }

        // Same consonants, vowels written differently.
        if (fuzzyOk) {
            forEachWithPrefix(curatedByVskel, qv, { curated[it].vskel }) { row, skel ->
                val c = curated[row]
                if (skel == qv) add(c.khmer, 2, 0, c.gloss)
                else add(c.khmer, 3, (skel.length - qv.length).toLong(), c.gloss)
                true
            }
        }

        // The 62k lexicon, matched on how the word sounds rather than how it is
        // spelled. `sub` keeps corpus rank as the tie-break, so common words win.
        if (fuzzyOk && lexKhmer.isNotEmpty()) {
            val qk = qskel(q)
            if (qk.length >= 2) {
                var visited = 0
                forEachWithPrefix(lexBySkel, qk, { lexSkel[it] }) { row, skel ->
                    if (skel == qk) add(lexKhmer[row], 4, row.toLong())
                    else add(lexKhmer[row], 5, (skel.length - qk.length) * 100_000L + row)
                    ++visited < MAX_LEXICON_VISITS
                }
            }
        }

        return hits.entries
            .map { (khmer, hit) ->
                val uses = learned.uses(khmer)
                val boost = if (uses >= 4) 2 else if (uses >= 1) 1 else 0
                Triple(khmer, hit, (hit.tier - boost).toLong() * 1_000_000_000_000L + hit.sub - uses)
            }
            .sortedBy { it.third }
            .take(limit)
            .map { Candidate(it.first, it.second.gloss) }
    }

    companion object {
        private const val MAX_TOKEN = 32
        // A guard against a one-letter skeleton matching half the lexicon. The
        // ranking still runs over everything visited, so raising this costs
        // time but never quality.
        private const val MAX_LEXICON_VISITS = 4000
        private const val GLOSS_YOURS = "yours"

        /** Collapse vowel runs, so "sousdey" and "suosdei" key alike. */
        fun vskel(s: String): String = s.replace(VOWEL_RUN, "8")

        private val VOWEL_RUN = Regex("[aeiouy]+")

        private val DIGRAPH = mapOf(
            "chh" to "C", "ch" to "C", "kh" to "K", "gh" to "K", "th" to "T",
            "dh" to "T", "dd" to "T", "ph" to "P", "bh" to "P", "ng" to "G",
            "nh" to "J", "ny" to "J", "gn" to "J",
        )
        private val SINGLE = mapOf(
            'k' to "K", 'g' to "K", 'c' to "C", 'j' to "C", 't' to "T", 'd' to "T",
            'b' to "P", 'p' to "P", 'f' to "P", 's' to "S", 'h' to "H", 'm' to "M",
            'n' to "N", 'l' to "L", 'x' to "S", 'z' to "S",
        )

        private fun isVowel(c: Char?) = c != null && c in "aeiou"

        /**
         * The same coarse sound-class skeleton the lexicon was built with:
         * consonant classes plus V for any vowel run, repeats collapsed. This is
         * what lets a sloppy spelling still land on the right word.
         */
        fun qskel(query: String): String {
            val q = query.lowercase().filter { it in 'a'..'z' }
            val out = StringBuilder()
            fun push(c: String) {
                if (c.isNotEmpty() && (out.isEmpty() || !out.endsWith(c))) out.append(c)
            }

            var i = 0
            while (i < q.length) {
                val tri = if (i + 3 <= q.length) q.substring(i, i + 3) else ""
                val di = if (i + 2 <= q.length) q.substring(i, i + 2) else ""
                val ch = q[i]
                val next = q.getOrNull(i + 1)
                val prev = q.getOrNull(i - 1)

                val triClass = DIGRAPH[tri]
                if (triClass != null) { push(triClass); i += 3; continue }
                val diClass = DIGRAPH[di]
                if (diClass != null) { push(diClass); i += 2; continue }

                if (isVowel(ch)) { push("V"); i++; continue }

                if (ch == 'w' || ch == 'v' || ch == 'y') {
                    // An onset is a consonant; a coda merges into the vowel.
                    if (isVowel(next)) push(if (ch == 'y') "Y" else "W") else push("V")
                    i++
                    continue
                }
                if (ch == 'r') {
                    // Khmer r is only pronounced before a vowel.
                    if (isVowel(next)) push("R")
                    i++
                    continue
                }
                if (ch == 's' && isVowel(prev) && !isVowel(next)) {
                    push("H")  // coda s sounds like h: monus = monuh
                    i++
                    continue
                }
                if (ch == 'q') { i++; continue }  // glottal stop, usually not typed

                push(SINGLE[ch] ?: "")
                i++
            }
            return out.toString()
        }
    }
}

File: ios/Shared/Suggestions.swift
//
//  The word suggestions that make this keyboard worth using: you spell a word
//  roughly as it sounds and the right Khmer comes back, the way a Pinyin IME
//  works.
//
//  Two lexicons, searched in tiers, ported from dict.js:
//
//    curated.tsv  ~440 common words, each listed under every romanization the
//                 web app's reverse-transliterator derives for it. Those keys
//                 are precomputed by tools/build_android.mjs, so the
//                 transliterator itself never had to be ported and its output
//                 cannot come out different from the browser's.
//    words.txt    62,000 words from the Google Khmer pronunciation lexicon,
//                 stored as coarse sound-class skeletons and ordered by corpus
//                 frequency, so line number is rank.
//
//  Tiers, lowest wins: personal(-1), curated exact(0), curated prefix(1),
//  curated vowel-skeleton(2, 3), lexicon skeleton exact(4), prefix(5).
//
//  Lookups binary-search sorted indices rather than scanning. A keyboard
//  extension gets a fraction of a normal app's CPU and memory, and scanning
//  77,000 entries per keystroke on the main thread both delays the word bar and
//  makes the keyboard miss taps.
//

import Foundation

struct Candidate: Equatable {
    let khmer: String
    let gloss: String

    init(_ khmer: String, gloss: String = "") {
        self.khmer = khmer
        self.gloss = gloss
    }
}

/// What the user has picked before, so their words float up.
protocol Learning {
    func uses(_ khmer: String) -> Int
    var personal: [String: String] { get }
}

struct NoLearning: Learning {
    func uses(_ khmer: String) -> Int { 0 }
    var personal: [String: String] { [:] }
}

final class Suggestions {

    private struct Curated {
        let key: String
        let vskel: String
        let khmer: String
        let gloss: String
    }

    private var curated: [Curated] = []
    private var lexKhmer: [String] = []
    private var lexSkel: [String] = []

    private var curatedByKey: [Int] = []
    private var curatedByVskel: [Int] = []
    private var lexBySkel: [Int] = []

    private let learned: Learning

    var lexiconSize: Int { lexKhmer.count }

    init(curatedTSV: String, lexiconTSV: String, learned: Learning = NoLearning()) {
        self.learned = learned

        curated.reserveCapacity(16_000)
        for line in curatedTSV.split(separator: "\n", omittingEmptySubsequences: true) {
            if line.hasPrefix("#") { continue }
            let parts = line.split(separator: "\t", omittingEmptySubsequences: false)
            if parts.count >= 3 {
                curated.append(
                    Curated(
                        key: String(parts[0]),
                        vskel: String(parts[1]),
                        khmer: String(parts[2]),
                        gloss: parts.count > 3 ? String(parts[3]) : ""
                    )
                )
            }
        }

        lexKhmer.reserveCapacity(62_000)
        lexSkel.reserveCapacity(62_000)
        for line in lexiconTSV.split(separator: "\n", omittingEmptySubsequences: true) {
            if line.hasPrefix("#") { continue }
            guard let tab = line.firstIndex(of: "\t") else { continue }
            lexKhmer.append(String(line[line.startIndex..<tab]))
            lexSkel.append(String(line[line.index(after: tab)...]).trimmingCharacters(in: .whitespaces))
        }

        curatedByKey = Self.sortedIndices(curated.count) { self.curated[$0].key }
        curatedByVskel = Self.sortedIndices(curated.count) { self.curated[$0].vskel }
        lexBySkel = Self.sortedIndices(lexSkel.count) { self.lexSkel[$0] }
    }

    private static func sortedIndices(_ count: Int, _ key: (Int) -> String) -> [Int] {
        Array(0..<count).sorted { key($0) < key($1) }
    }

    /// First position in `order` whose string is >= `target`.
    private func lowerBound(_ order: [Int], _ target: String, _ key: (Int) -> String) -> Int {
        var lo = 0
        var hi = order.count
        while lo < hi {
            let mid = (lo + hi) / 2
            if key(order[mid]) < target { lo = mid + 1 } else { hi = mid }
        }
        return lo
    }

    /// Visit every row whose string starts with `prefix`. `visit` returns false to stop.
    private func forEachWithPrefix(
        _ order: [Int],
        _ prefix: String,
        _ key: (Int) -> String,
        _ visit: (Int, String) -> Bool
    ) {
        var i = lowerBound(order, prefix, key)
        while i < order.count {
            let row = order[i]
            let value = key(row)
            if !value.hasPrefix(prefix) { return }
            if !visit(row, value) { return }
            i += 1
        }
    }

    private struct Hit {
        var tier: Int
        var sub: Int
        var gloss: String
    }

    func suggest(_ rawQuery: String, limit: Int = 8) -> [Candidate] {
        var q = rawQuery.lowercased().trimmingCharacters(in: .whitespaces)
        if q.isEmpty { return [] }
        // Bound the work: a long paste should never hang the keyboard.
        if q.count > Self.maxToken { q = String(q.suffix(Self.maxToken)) }

        var hits: [String: Hit] = [:]
        func add(_ khmer: String, _ tier: Int, _ sub: Int, _ gloss: String = "") {
            if var existing = hits[khmer] {
                if tier < existing.tier || (tier == existing.tier && sub < existing.sub) {
                    existing.tier = tier
                    existing.sub = sub
                }
                if existing.gloss.isEmpty && !gloss.isEmpty { existing.gloss = gloss }
                hits[khmer] = existing
            } else {
                hits[khmer] = Hit(tier: tier, sub: sub, gloss: gloss)
            }
        }

        for (spelling, khmer) in learned.personal {
            if spelling == q {
                add(khmer, -1, 0, Self.glossYours)
            } else if spelling.hasPrefix(q) {
                add(khmer, -1, spelling.count - q.count, Self.glossYours)
            }
        }

        let qv = Self.vskel(q)
        let fuzzyOK = q.count >= 3

        // Spelled exactly, or the query is the start of a longer spelling.
        forEachWithPrefix(curatedByKey, q, { self.curated[$0].key }) { row, key in
            let c = self.curated[row]
            if key == q { add(c.khmer, 0, 0, c.gloss) }
            else { add(c.khmer, 1, key.count - q.count, c.gloss) }
            return true
        }

        // Same consonants, vowels written differently.
        if fuzzyOK {
            forEachWithPrefix(curatedByVskel, qv, { self.curated[$0].vskel }) { row, skel in
                let c = self.curated[row]
                if skel == qv { add(c.khmer, 2, 0, c.gloss) }
                else { add(c.khmer, 3, skel.count - qv.count, c.gloss) }
                return true
            }
        }

        // The 62k lexicon, matched on how the word sounds rather than how it is
        // spelled. `sub` keeps corpus rank as the tie-break, so common words win.
        if fuzzyOK && !lexKhmer.isEmpty {
            let qk = Self.qskel(q)
            if qk.count >= 2 {
                var visited = 0
                forEachWithPrefix(lexBySkel, qk, { self.lexSkel[$0] }) { row, skel in
                    if skel == qk { add(self.lexKhmer[row], 4, row) }
                    else { add(self.lexKhmer[row], 5, (skel.count - qk.count) * 100_000 + row) }
                    visited += 1
                    return visited < Self.maxLexiconVisits
                }
            }
        }

        return hits
            .map { khmer, hit -> (String, Hit, Int) in
                let uses = learned.uses(khmer)
                let boost = uses >= 4 ? 2 : (uses >= 1 ? 1 : 0)
                return (khmer, hit, (hit.tier - boost) * 1_000_000_000 + hit.sub - uses)
            }
            .sorted { $0.2 < $1.2 }
            .prefix(limit)
            .map { Candidate($0.0, gloss: $0.1.gloss) }
    }

    // MARK: - Sound skeletons

    private static let maxToken = 32
    /// A guard against a short skeleton matching half the lexicon. Ranking still
    /// runs over everything visited, so raising this costs time, never quality.
    private static let maxLexiconVisits = 4000
    private static let glossYours = "yours"

    /// Collapse vowel runs, so "sousdey" and "suosdei" key alike.
    static func vskel(_ s: String) -> String {
        var out = ""
        var inRun = false
        for ch in s {
            if "aeiouy".contains(ch) {
                if !inRun { out.append("8"); inRun = true }
            } else {
                out.append(ch)
                inRun = false
            }
        }
        return out
    }

    private static let digraphs: [String: String] = [
        "chh": "C", "ch": "C", "kh": "K", "gh": "K", "th": "T",
        "dh": "T", "dd": "T", "ph": "P", "bh": "P", "ng": "G",
        "nh": "J", "ny": "J", "gn": "J",
    ]
    private static let singles: [Character: String] = [
        "k": "K", "g": "K", "c": "C", "j": "C", "t": "T", "d": "T",
        "b": "P", "p": "P", "f": "P", "s": "S", "h": "H", "m": "M",
        "n": "N", "l": "L", "x": "S", "z": "S",
    ]

    private static func isVowel(_ c: Character?) -> Bool {
        guard let c else { return false }
        return "aeiou".contains(c)
    }

    /// The same coarse sound-class skeleton the lexicon was built with:
    /// consonant classes plus V for any vowel run, repeats collapsed. This is
    /// what lets a sloppy spelling still land on the right word.
    static func qskel(_ query: String) -> String {
        let q = Array(query.lowercased().filter { $0.isASCII && $0.isLetter })
        var out: [String] = []
        func push(_ c: String) {
            if !c.isEmpty && out.last != c { out.append(c) }
        }

        var i = 0
        while i < q.count {
            let ch = q[i]
            let next: Character? = i + 1 < q.count ? q[i + 1] : nil
            let prev: Character? = i > 0 ? q[i - 1] : nil

            if i + 3 <= q.count, let cls = digraphs[String(q[i..<(i + 3)])] {
                push(cls); i += 3; continue
            }
            if i + 2 <= q.count, let cls = digraphs[String(q[i..<(i + 2)])] {
                push(cls); i += 2; continue
            }
            if isVowel(ch) { push("V"); i += 1; continue }

            if ch == "w" || ch == "v" || ch == "y" {
                // An onset is a consonant; a coda merges into the vowel.
                push(isVowel(next) ? (ch == "y" ? "Y" : "W") : "V")
                i += 1
                continue
            }
            if ch == "r" {
                // Khmer r is only pronounced before a vowel.
                if isVowel(next) { push("R") }
                i += 1
                continue
            }
            if ch == "s", isVowel(prev), !isVowel(next) {
                push("H")  // coda s sounds like h: monus = monuh
                i += 1
                continue
            }
            if ch == "q" { i += 1; continue }  // glottal stop, usually not typed

            push(singles[ch] ?? "")
            i += 1
        }
        return out.joined()
    }
}
